soda/core/soda/execution/data_source.py

```diff
@@ -205,33 +205,73 @@ def sql_select_all(self, table_name: str, limit: int | None = None) -> str:
     # For a table, get the columns metadata
     ############################################
 
-    def get_table_columns(self, table_name: str, query_name: str) -> dict[str, str]:
+    def get_table_columns(
+        self,
+        table_name: str,
+        query_name: str,
+        included_columns: list[str] | None = None,
+        excluded_columns: list[str] | None = None,
+    ) -> dict[str, str] | None:
         """
         :return: A dict mapping column names to data source data types.  Like eg
         {"id": "varchar", "size": "int8", ...}
         """
         query = Query(
             data_source_scan=self.data_source_scan,
             unqualified_query_name=query_name,
-            sql=self.sql_get_table_columns(table_name),
+            sql=self.sql_get_table_columns(
+                table_name, included_columns=included_columns, excluded_columns=excluded_columns
+            ),
         )
         query.execute()
-        return {row[0]: row[1] for row in query.rows}
+        if len(query.rows) > 0:
+            return {row[0]: row[1] for row in query.rows}
+        return None
 
     def create_table_columns_query(self, partition: Partition, schema_metric: SchemaMetric) -> TableColumnsQuery:
         return TableColumnsQuery(partition, schema_metric)
 
-    def sql_get_table_columns(self, table_name: str) -> str:
+    def sql_get_table_columns(
+        self, table_name: str, included_columns: list[str] | None = None, excluded_columns: list[str] | None = None
+    ) -> str:
+        # build optional filter clauses
+        if self.database:
+            database_filter = f" \n  AND lower({self.column_metadata_catalog_column()}) = '{self.database.lower()}'"
```

Comment:
Elsewhere I take a different approach that I like better:

filter_clauses = []

...append to the filterclauses like eg...
filter_clauses.append(f"lower({self.column_metadata_catalog_column()}) = '{self.database.lower()}'")

then at the end
filter_clauses_sql = "\n    AND ".join(filter_clauses)

then in the query template something like:
```
return (
    f"SELECT ... \n"
    f"FROM ... \n"
    f"WHERE {filter_clauses_sql} \n"
    ...
)
```

That way you also prevent if-then-else in case you only have 1 clause.  And the approach can be consistent.

Reply:
What I find potentially annoying with the other approaches or the concatenation approaches is that they are not explicit about how things land in the template.

In this approach, I see the template at then end. I know which elements I potentially have and what they all do via explicit names.

I don't really have an issue with having a few if here and there as they actually make things quite explicit. I feel like maintaining a piece of code that must concatenate or inject in the string such as

`sql +=` which was done before much more damaging to maintain since the layout of the code itself has to happen in the right order whereas here the template at the end takes care of that.

Now if you think you don't want several placeholders in the filter clauses (not sure why that would be because I think they document nicely what kind of things we filter on) and want instead a bunch of strings to be appended into a list, we'll actually still need a fair bit of if and else's.

Also, note that the if-else logic for `self.database` and `self.schema` were already there as if else's so it doesn't look like we have strong consistency.

To me this comes down to stylistic preference, and I am not convinced that the alternative is fundamentally better. If there is a reason other than that that could make maintenance harder (which I don't think it does) or potentially create bugs, I'll be happy to change this but at this point, I would say that the unifying query building style should be in an effort that is focused on that and across the board rather than the scope of this PR.

Reply:
agreed it's style preference.  mostly i wanted to share the join-approach in case you would like it better.